ui/decredmaterial/switch.go

```diff
@@ -54,6 +67,106 @@ func (t *Theme) SwitchButtonText(i []SwitchItem) *SwitchButtonText {
 	return sw
 }
 
+func (s *Switch) Layout(gtx layout.Context) layout.Dimensions {
+	trackWidth := gtx.Px(unit.Dp(32))
+	trackHeight := gtx.Px(unit.Dp(20))
+	thumbSize := gtx.Px(unit.Dp(18))
+	trackOff := float32(thumbSize-trackHeight) * .5
+
+	// Draw track.
+	stack := op.Save(gtx.Ops)
+	trackCorner := float32(trackHeight) / 2
+	trackRect := f32.Rectangle{Max: f32.Point{
+		X: float32(trackWidth),
+		Y: float32(trackHeight),
+	}}
+
+	col := s.inactive
+	if s.disabled {
+		col = Disabled(s.inactive)
+		s.thumbColor = Disabled(s.thumbColor)
+	}
+	if s.IsChecked() {
+		col = s.active
+	}
+
+	trackColor := col
+	op.Offset(f32.Point{Y: trackOff}).Add(gtx.Ops)
+	clip.UniformRRect(trackRect, trackCorner).Add(gtx.Ops)
+	paint.ColorOp{Color: trackColor}.Add(gtx.Ops)
+	paint.PaintOp{}.Add(gtx.Ops)
+	stack.Load()
+
+	// Compute thumb offset and color.
+	stack = op.Save(gtx.Ops)
+	if s.IsChecked() {
+		off := trackWidth - thumbSize
+		op.Offset(f32.Point{X: float32(off)}).Add(gtx.Ops)
+	}
+
+	thumbRadius := float32(thumbSize) / 2
+
+	// Draw thumb shadow, a translucent disc slightly larger than the
+	// thumb itself.
+	// Center shadow horizontally and slightly adjust its Y.
+	paint.FillShape(gtx.Ops, col,
+		clip.Circle{
+			Center: f32.Point{X: thumbRadius, Y: thumbRadius + .25},
+			Radius: thumbRadius + 1,
+		}.Op(gtx.Ops))
+
+	// Draw thumb.
+	paint.FillShape(gtx.Ops, s.thumbColor,
+		clip.Circle{
+			Center: f32.Point{X: thumbRadius, Y: thumbRadius},
+			Radius: thumbRadius,
+		}.Op(gtx.Ops))
+
+	// Set up click area.
+	stack = op.Save(gtx.Ops)
+	clickSize := gtx.Px(unit.Dp(40))
+	clickOff := f32.Point{
+		X: (float32(trackWidth) - float32(clickSize)) * .5,
+		Y: (float32(trackHeight)-float32(clickSize))*.5 + trackOff,
+	}
+	op.Offset(clickOff).Add(gtx.Ops)
+	sz := image.Pt(clickSize, clickSize)
+	pointer.Ellipse(image.Rectangle{Max: sz}).Add(gtx.Ops)
+	gtx.Constraints.Min = sz
+	s.switchWidget.Layout(gtx)
+	stack.Load()
+
+	dims := image.Point{X: trackWidth, Y: thumbSize}
+	return layout.Dimensions{Size: dims}
+}
+
+func (s *Switch) Changed() bool {
+	return s.switchWidget.Changed()
+}
+
+func (s *Switch) IsChecked() bool {
+	if s.disabled {
+		s.SetChecked(false)
```

Comment:
Why is this value set in a read-only function? Also it's possible for a switch to be checked while in a disabled state. The disabled state simply means no user interaction is allowed.